contracts/pcv/balancer/BalancerPool2Lens.sol

```diff
@@ -0,0 +1,178 @@
+
+import "./IVault.sol";
+import "./IWeightedPool.sol";
+import "../../external/gyro/ExtendedMath.sol";
+import "../IPCVDepositBalances.sol";
+import "../../oracle/IOracle.sol";
+import "../../Constants.sol";
+
+/// @title BalancerPool2Lens
+/// @author Fei Protocol
+/// @notice a contract to read tokens & fei out of a contract that reports balance in Balancer LP tokens.
+/// Limited to BPTs that have 2 underlying tokens.
+/// @notice this contract use code similar to BPTLens (that reads a whole pool).
+contract BalancerPool2Lens is IPCVDepositBalances {
+    using ExtendedMath for uint256;
+
+    /// @notice FEI token address
+    address private constant FEI = 0x956F47F50A910163D8BF957Cf5846D573E7f87CA;
+
+    /// @notice the deposit inspected
+    address public immutable depositAddress;
+
+    /// @notice the token the lens reports balances in
+    address public immutable override balanceReportedIn;
+
+    /// @notice the balancer pool to look at
+    IWeightedPool public immutable pool;
+
+    /// @notice the Balancer V2 Vault
+    IVault public immutable balancerVault;
+
+    // the pool id on balancer
+    bytes32 internal immutable id;
+
+    // the index of balanceReportedIn on the pool
+    uint256 internal immutable index;
+
+    /// @notice true if FEI is in the pair
+    bool public immutable feiInPair;
+
+    /// @notice true if FEI is the reported balance
+    bool public immutable feiIsReportedIn;
+
+    /// @notice the oracle for balanceReportedIn token
+    IOracle public immutable reportedOracle;
+
+    /// @notice the oracle for the other token in the pair (not balanceReportedIn)
+    IOracle public immutable otherOracle;
+
+    constructor(
+        address _depositAddress,
+        address _token,
+        IWeightedPool _pool,
+        IOracle _reportedOracle,
+        IOracle _otherOracle,
+        bool _feiIsReportedIn,
+        bool _feiIsOther
+    ) {
+        depositAddress = _depositAddress;
+
+        pool = _pool;
+        IVault _vault = _pool.getVault();
+        balancerVault = _vault;
+
+        bytes32 _id = _pool.getPoolId();
+        id = _id;
+        (IERC20[] memory tokens, , ) = _vault.getPoolTokens(_id);
+
+        // Check the token is in the BPT and its only a 2 token pool
+        require(address(tokens[0]) == _token || address(tokens[1]) == _token);
+        require(tokens.length == 2);
+        balanceReportedIn = _token;
+
+        index = address(tokens[0]) == _token ? 0 : 1;
+
+        feiIsReportedIn = _feiIsReportedIn;
+        feiInPair = _feiIsReportedIn || _feiIsOther;
+
+        reportedOracle = _reportedOracle;
+        otherOracle = _otherOracle;
+    }
+
+    function balance() public view override returns (uint256) {
+        (, uint256[] memory balances, ) = balancerVault.getPoolTokens(id);
+        uint256 bptsOwned = IPCVDepositBalances(depositAddress).balance();
+        uint256 totalSupply = pool.totalSupply();
+
+        return (balances[index] * bptsOwned) / totalSupply;
+    }
```

Comment:
this function should return underlying reserves not bpts, right?

Reply:
right, it returns `balances[index]` which is an underlying token balance, multiplied by the % of the pool owned by another contract (that return its balance in BPTs).

Example:
- Pool has 100 BPTs, [1000 tokenA, 5000 tokenB]
- A PCVDeposit reports that it has 50 BPTs
- This lens, reporting the `tokenA` balance of the PCVDeposit, reads the reserves (1000) and multiplies by 50 (the owned BPTs) divided by the total supply (100) => 500

Reply:
ah right this doesn't need to be manipulation resistant. Makes sense